File: basic_java/src/test/java/pers/xiaoming/spark/core/TopNSelectorTest.java
package pers.xiaoming.spark.core;

import org.junit.AfterClass;
import org.junit.Assert;
import org.junit.BeforeClass;
import org.junit.Test;
import pers.xiaoming.spark.SparkCoreDemoTestBase;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

public class TopNSelectorTest extends SparkCoreDemoTestBase {
    private static final int N = 5;
    private static TopNSelector topNSelector;

    private static List<Integer> testInput;
    private static List<Integer> expectedResult;

    @BeforeClass
    public static void setupTestData() {
        int start = 0;
        int end = 100;

        topNSelector = new TopNSelector(sc);
        testInput = new ArrayList<>(100);

        for (int i = start; i <= end; i++) {
            testInput.add(i);
        }

        expectedResult = new ArrayList<>(N);

        for (int i = end; i > end - N; i--) {
            expectedResult.add(i);
        }
    }

    @AfterClass
    public static void close() {
        try {
            topNSelector.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    @Test
    public void topNSortImplTest() {
        List<Integer> result = topNSelector.getTopNSortImpl(testInput, N);
        Assert.assertEquals(expectedResult, result);
    }

    @Test
    public void topNHeapImplTest() {
        List<Integer> result = topNSelector.getTopNHeapImpl(testInput, N);
        result.sort(Comparator.comparingInt(i -> -i));
        Assert.assertEquals(expectedResult, result);
    }
}

File: basic_scala/src/test/scala/pers/xiaoming/spark/core/TopNSelectorTest.scala
package pers.xiaoming.spark.core

import org.junit.{Assert, BeforeClass, Ignore, Test}
import pers.xiaoming.spark.SparkCoreDemoTestBase

@Ignore
class TopNSelectorTest extends SparkCoreDemoTestBase {
  private val N = 5
  private val topNSelector = new TopNSelector(SparkCoreDemoTestBase.sc)
  private val input = List(1 to 100: _*)
  private val expectResult = List(100, 99, 98, 97, 96)

  // Runtime: 1.56s
  @Test
  def topNSortImplTest: Unit = {
    Assert.assertEquals(expectResult, topNSelector.getTopNSortImpl[Int](input, N))
  }

  // Runtime: 40ms
  @Test
  def topNSortHeapTest: Unit = {
    Assert.assertEquals(expectResult, topNSelector.getTopNHeapImpl[Int](input, N).sorted(Ordering[Int].reverse))
  }
}
